python/tvm/topi/cuda/nms.py

```diff
@@ -519,14 +557,90 @@ def non_max_suppression(
     if return_indices:
-        return box_indices
+        out_shape = box_indices.shape
+        valid_box_count_shape = [box_indices.shape[0], 1]
+        valid_box_count = tvm.tir.decl_buffer(valid_box_count_shape, "int32", "valid_box_count")
+        output = tvm.tir.decl_buffer(box_indices.shape, "int32", "output")
+        return te.extern(
+            [out_shape, valid_box_count_shape],
+            [box_indices],
+            lambda ins, outs: rearrange_indices_out_ir(ins[0], outs[0], outs[1]),
+            dtype="int32",
+            out_buffers=[output, valid_box_count],
+            name="rearrange_indices_out_gpu",
+            tag="rearrange_indices_out_gpu",
+        )
 
     return out
+
+
+def rearrange_indices_out_ir(data, output, valid_box_count):
+    """Hybrid routine to rearrange nms output to
+    move all valid entries to top.
+
+    Parameters
+    ----------
+    data : tvm.te.Tensor or numpy NDArray
+        NMS output. 3-D tensor with shape
+        [batch_size, num_anchors, 6] or
+        [batch_size, num_anchors, 5], or 2-D
+        tensor with shape [batch_size, num_anchors].
+
+    one: tvm.tir.const
+        Constant one with the same dtype as data.
+
+    batch_size: tvm.tir.IntImm or tvm.tir.Var
+        Batch size. We need to pass it in since hybrid script doesn't support
+        binding variable to symbolic dim.
+
+    num_anchors: tvm.tir.IntImm or tvm.tir.Var
+        Number of anchors.
+
+    Returns
+    -------
+    output : tvm.te.Tensor or numpy NDArray
+        2-D tensor with shape [batch_size, num_anchors].
+
+    valid_box_count : tvm.te.Tensor or numpy NDArray
+        Tensor with shape [batch_size, 1], indicates
+        the valid number of boxes.
+    """
+    batch_size = data.shape[0]
+    num_anchors = data.shape[1]
+
+    ib = tvm.tir.ir_builder.create()
+
+    data = ib.buffer_ptr(data)
+    valid_box_count = ib.buffer_ptr(valid_box_count)
+    output = ib.buffer_ptr(output)
+
+    with ib.new_scope():
+        i = te.thread_axis("blockIdx.x")
+        ib.scope_attr(i, "thread_extent", batch_size)
+        valid_idx = ib.allocate("int32", (1), name="valid_idx", scope="local")
+        valid_idx[0] = 0
+        with ib.for_range(0, num_anchors, name="j") as j:
+            with ib.if_scope(data[i, j] >= 0):
+                with ib.if_scope(data[i, j] > num_anchors):
+                    output[i, valid_idx[0]] = 0
+                    valid_idx[0] = valid_idx[0] + 1
+                with ib.else_scope():
+                    output[i, valid_idx[0]] = data[i, j]
+                    valid_idx[0] = valid_idx[0] + 1
+            with ib.else_scope():
+                with ib.if_scope(data[i, j] < -num_anchors):
+                    output[i, valid_idx[0]] = 0
+                    valid_idx[0] = valid_idx[0] + 1
+            with ib.if_scope(j >= valid_idx[0]):
+                output[i, j] = -1
+        valid_box_count[i, 0] = valid_idx[0]
+
+    return ib.get()
```

Comment:
Could you show the performance benchmark on some popular OD model workloads with the modified nms.py?

Reply:
I've been attempted to get SSD-Mobilenet from the ONNX model zoo working, but I"m hitting other bugs. I'll post some perf metrics as soon as I can get that working.

Reply:
I fixed a number of bugs in #6906 that allow me to run SSD-RN50 on CPU, but with those fixes, I'm hitting the isuse of cuda topk not supporting dynamic shapes, and the output of NMS is intrinsically dynamic. I'm back to trying to solve that problem before I can give you perf numbers.

Reply:
I'm still struggling to get full object detection models working on GPU due to the issues with dynamic topk in cuda, but while I work on that, I did some profiling of the unit test. With batch size = 2, num_anchors = 5, and num_classes = 2, I get the following performance on my 1070Ti:

Total Execution time: 900 microseconds
NMS time: 66 microseconds
get_valid_counts time: 15 microseconds.

it looks like the performance issues in the onnx kernel are all about the slicing/expand dims and loop bounds checking, here are the most expensive ops:

```
0                      fused_dyn_strided_slice_1        89.28450                                     
1                      fused_add_2                      87.05910                                     
2                      fused_less_min                   83.04660                                     
3                    fused_vision_non_max_suppression   66.43930                                     
4   fused_expand_dims_cast_concatenate_expand_dims...   53.65090                                     
5                      fused_reshape_concatenate        46.64780                                     
6                      fused_dyn_strided_slice          46.59800                                     
7                      fused_concatenate_2              33.06690                                     
8                      fused_dyn_broadcast_to_1         29.49310                                     
9                      fused_squeeze_concatenate        26.83250                                     
10                 fused_cast_expand_dims_concatenate   24.00320                                     
11  fused_concatenate_cast_like_less_cast_like_add...   23.53060                                     
12                     fused_dyn_strided_slice_2        23.39070                                     
13  fused_concatenate_cast_like_less_cast_like_add...   22.79010                                     
14                     fused_shape_of_1                 21.37450                                     
15                     fused_shape_of_2                 21.02070                                     
16                     fused_dyn_broadcast_to           18.43470                                     
17                     fused_vision_get_valid_counts    14.89870
```

cc @jwfromm @tkonolige

Reply:
Could you leverage more in the changes for this file?

Reply:
I have not been able to pass tests with the kernels currently in master, so I can't compare apples to apples